localstack/aws/protocol/router.py

```diff
@@ -0,0 +1,234 @@
+import re
+from collections import defaultdict
+from typing import Any, Dict, List, Mapping, NamedTuple, Optional, Tuple
+from urllib.parse import parse_qs, unquote
+
+from botocore.model import OperationModel, ServiceModel, StructureShape
+from werkzeug.datastructures import Headers, MultiDict
+from werkzeug.exceptions import NotFound
+from werkzeug.routing import Map, MapAdapter, Rule
+
+from localstack.http import Request
+from localstack.http.request import get_raw_path
+
+
+class _RequiredArgsRule:
+    """
+    Specific Rule implementation which checks if a set of certain required header and query parameters are matched by
+    a specific request.
+    """
+
+    endpoint: Any
+    required_query_args: Optional[Mapping[str, List[Any]]]
+    required_header_args: List[str]
+    match_score: int
+
+    def __init__(
+        self,
+        endpoint: Any,
+        required_query_args: Optional[Mapping[str, List[Any]]],
+        required_header_args: Optional[List[str]],
+    ) -> None:
+        super().__init__()
+        self.endpoint = endpoint
+        self.required_query_args = required_query_args or {}
+        self.required_header_args = required_header_args or []
+        self.match_score = 1 + len(self.required_query_args) + len(self.required_header_args)
+
+    def matches(self, query_args: "MultiDict[str, str]", headers: Headers) -> bool:
+        """
+        Returns true if the given query args and the given headers of a request match the required query args and
+        headers of this rule.
+        :param query_args: query arguments of the incoming request
+        :param headers: headers of the incoming request
+        :return: True if the query args and headers match the required args of this rule
+        """
+        if self.required_query_args:
+            for key, values in self.required_query_args.items():
+                if key not in query_args:
+                    return False
+                # if a required query arg also has a list of required values set, the values need to match as well
+                if values:
+                    query_arg_values = query_args.getlist(key)
+                    for value in values:
+                        if value not in query_arg_values:
+                            return False
+
+        if self.required_header_args:
+            for key in self.required_header_args:
+                if key not in headers:
+                    return False
+
+        return True
+
+
+class _HttpOperation(NamedTuple):
+    operation: OperationModel
+    path: str
+    method: str
+    query_args: Mapping[str, List[str]]
+    header_args: List[str]
+
+    @staticmethod
+    def from_operation(op: OperationModel) -> "_HttpOperation":
+        """Useful intermediary representation of the 'http' block of an operation to make code cleaner"""
```

Comment:
nit: maybe move this comment up to the class doc, and add some concrete method doc here on how it works